util/genrest/resttools/headers.go

```diff
@@ -15,19 +15,70 @@
 package resttools
 
 import (
+	"fmt"
 	"net/http"
+	"strings"
 )
 
 const (
 	headerNameContentType      = "Content-Type"
 	headerValueContentTypeJSON = "application/json"
+
+	headerNameAPIClient            = "X-Goog-Api-Client"
+	headerValueTransportRESTPrefix = "rest/"
+	headerValueClientGAPICPrefix   = "gapic/"
 )
 
 // PopulateRequestHeaders inspects request and adds the correct headers. This
 // is useful for tests where we're not trying to send incorrect
 // headers.
 func PopulateRequestHeaders(request *http.Request) {
+	header := http.Header{}
+	header.Set(headerNameAPIClient, fmt.Sprintf("%s0.0.0 %s0.0.0", headerValueTransportRESTPrefix, headerValueClientGAPICPrefix))
+
 	if request.Body != nil {
-		request.Header = http.Header{headerNameContentType: []string{headerValueContentTypeJSON}}
+		header.Set(headerNameContentType, headerValueContentTypeJSON)
+	}
+
+	request.Header = header
+}
+
+// CheckContentType checks header to ensure the expected JSON content type is specified.
+func CheckContentType(header http.Header) error {
+	if content, ok := header[headerNameContentType]; !ok || len(content) != 1 || !strings.HasPrefix(strings.ToLower(strings.TrimSpace(content[0])), headerValueContentTypeJSON) {
+		return fmt.Errorf("(HeaderContentTypeError) did not find expected HTTP header %q: %q", headerNameContentType, headerValueContentTypeJSON)
+	}
+	return nil
+}
+
+// CheckAPIClientHeader verifies that the "x-goog-api-client" header contains the expected tokens
+// ("rest/..." and "gapic/...").
+func CheckAPIClientHeader(header http.Header) error {
+	content, ok := header[headerNameAPIClient]
+	if !ok || len(content) != 1 {
+		return fmt.Errorf("(HeaderAPIClientError) did not find expected HTTP header %q: %q", headerNameAPIClient, headerValueTransportRESTPrefix)
+	}
+
+	var haveREST, haveGAPIC bool
+	for _, token := range strings.Split(content[0], " ") {
+		trimmed := strings.TrimSpace(token)
+		if !haveREST && strings.HasPrefix(trimmed, headerValueTransportRESTPrefix) {
+			haveREST = true
+		} else if !haveGAPIC && strings.HasPrefix(trimmed, headerValueClientGAPICPrefix) {
+			haveGAPIC = true
+		} else {
+			// nothing changed
+			continue
```

Comment:
Is this else case necessary? I haven't thought through all of the scenarios, but thought I'd ask.

Reply:
Not necessary, but a pedantic optimization. We could have just three linear if-statements and things would work fine, if "inefficiently". But then I decided to be more explicit (which might be clearer or muddier, not sure): if `haveREST` is not set yet, see whether we can set it; otherwise if `haveGAPIC` is not set yet, see whether we can set it; otherwise, we didn't change either, so continue with the loop (this else case). Fall-through: one of the two variables was just set, so let's see whether we can exit the loop prematurely.

(The efficiency point is just a principle thing, not practical; we're only dealing with a handful of tokens in the header.)